db/version_set.cc

```diff
@@ -2758,24 +2758,43 @@ void VersionStorageInfo::ComputeCompactionScore(
           // Level-based involves L0->L0 compactions that can lead to oversized
           // L0 files. Take into account size as well to avoid later giant
           // compactions to the base level.
-          uint64_t l0_target_size = mutable_cf_options.max_bytes_for_level_base;
-          if (immutable_options.level_compaction_dynamic_level_bytes &&
-              level_multiplier_ != 0.0) {
-            // Prevent L0 to Lbase fanout from growing larger than
-            // `level_multiplier_`. This prevents us from getting stuck picking
-            // L0 forever even when it is hurting write-amp. That could happen
-            // in dynamic level compaction's write-burst mode where the base
-            // level's target size can grow to be enormous.
-            l0_target_size =
-                std::max(l0_target_size,
-                         static_cast<uint64_t>(level_max_bytes_[base_level_] /
-                                               level_multiplier_));
-          }
-          score =
-              std::max(score, static_cast<double>(total_size) / l0_target_size);
-          if (immutable_options.level_compaction_dynamic_level_bytes &&
-              score > 1.0) {
-            score *= kScoreScale;
+          // If score in L0 is always too high, L0->L1 will always be
+          // prioritized over L1->L2 compaction and L1 will accumulate to
+          // too large. But if L0 score isn't high enough, L0 will accumulate
+          // and data is not moved to L1 fast enough. With potential L0->L0
+          // compaction, number of L0 files aren't always an indication of
+          // L0 oversizing, and we also need to consider total size of L0.
+          if (immutable_options.level_compaction_dynamic_level_bytes) {
+            if (total_size >= mutable_cf_options.max_bytes_for_level_base) {
+              // When calculating estimated_compaction_needed_bytes, we assume
+              // L0 is qualified as pending compactions. We will need to make
+              // sure that it qualifies for compaction.
+              // It might be guafanteed by logic below anyway, but we are
+              // explicit here to make sure we don't stop writes with no
+              // compaction scheduled.
+              score = std::max(score, 1.01);
```

Comment:
Why not use `std::max(score, std::min(level_multiplier_, total_size / max_bytes_for_level_base))` so increased Size(L0) would actually cause score to increase even when it's smaller than Size(Lbase)? In the current PR, for example, if Score(Lbase) = 3, then Size(L0)/Size(Lbase) must be larger than 3 to prioritize L0->Lbase. That seems like too large an L0 and also an hourglass shape. With `total_size / max_bytes_for_level_base` we will at least prefer L0 whenever it's larger than Lbase.

Reply:
I tried that. I faced a problem that `total_size / max_bytes_for_level_base` can be very large. In the current algorithm, actually L1 will have a much smaller score than L0 (because we de-prioritize a level when there are lots of pending data in upper levels), and this causes data stuck in L1. That's why it is always divided by Size(Lbase) if Lbase is relatively large, this is to make it possible for L0 score could be smaller than Lbase if Lbase is large.

Reply:
> I faced a problem that total_size / max_bytes_for_level_base can be very large

The std::min(level_multiplier_, total_size / max_bytes_for_level_base) intended to prevent that case

> That's why it is always divided by Size(Lbase) if Lbase is relatively large, this is to make it possible for L0 score could be smaller than Lbase if Lbase is large.

I think Score(L0) < Score(Lbase) when Size(L0) < Size(Lbase) is good enough. The current formula makes Score(L0) < Score(Lbase) when Size(L0) < Score(Lbase) * Size(Lbase), which just seems like L0 may have to be much much bigger than Lbase in order for L0 to be prioritized.

Anyway I don't mind trying it if you want since hourglass LSM might work fine. I haven't measured it. So my accept still stands.

Reply:
> I think Score(L0) < Score(Lbase) when Size(L0) < Size(Lbase) is good enough. The current formula makes Score(L0) < Score(Lbase) when Size(L0) < Score(Lbase) * Size(Lbase), which just seems like L0 may have to be much much bigger than Lbase in order for L0 to be prioritized.
The benchmark I ran (the one Mark used) actually shows that L0->L1 still feels prioritized perhaps too much against L1->L2. If L0 is larger than L1, L0->L1 is always prioritized against L1->L2, because the score Math makes L1 score < 0.5 (5 after scaling). Similarly, if number of files are more than number of L0 file trigger, we almost always pick from L0. L1->L2 usually gets picked up with L1 is significantly larger than L0.

Here are some snapshots of the benchmark runs (level multiplier is 8 and base level size base is 64MB, exact parameters used by Mark):
```
  L0     12/10  321.94 MB 
  L4     23/23  368.26 MB 
  L5     11/0   176.77 MB 
  L6    112/4    1.50 GB  
  L7    798/23  11.42 GB  

  L0      8/6   218.41 MB
  L4     24/24  374.77 MB
  L5     20/0   317.81 MB
  L6    118/5    1.53 GB 
  L7    814/25  11.69 GB 

  L0     10/6   278.14 MB
  L4     20/20  312.10 MB
  L5     11/0   175.13 MB
  L6    119/4    1.55 GB 
  L7    830/18  11.96 GB 

  L0      9/6   307.93 MB
  L4     20/20  321.85 MB
  L5     12/0   193.10 MB
  L6    113/0    1.52 GB 
  L7    853/0   12.21 GB 

  L0      5/0   131.31 MB 
  L4     26/0   402.99 MB 
  L5     18/0   282.12 MB 
  L6    123/3    1.60 GB  
  L7    871/15  12.42 GB  
```

Reply:
> If L0 is larger than L1, L0->L1 is always prioritized against L1->L2, because the score Math makes L1 score < 0.5 (5 after scaling).

Got it, did the math and agreed Score(L0) > Score(Lbase) when Size(L0) > Size(Lbase). I had forgotten to incorporate Size(L0) in the denominator of Score(Lbase) earlier (courtesy of `total_downcompact_bytes`). This is challenging to reason about but I am glad it works.

> Similarly, if number of files are more than number of L0 file trigger, we almost always pick from L0.

Understood this part already. The size-based component of Score(L0) is not used for this case.

Reply:
> the score Math makes L1 score < 0.5 (5 after scaling).

My calculation for Score(Lbase) is 10 * S'(Lbase) / (max_bytes_for_level_base + S(L0)). S'(Li) is the sum of compensated size of files not being compacted on level i, and S(Li) is sum of all actual file sizes on level i.

I don't see why S'(L0) > S'(Lbase) (assuming that's what you mean by L0 is larger than L1) means Score(Lbase) is < 5. It looks to be typically less than 10, unless Lbase files have high size compensation.

For my attempt to prove Score(L0) > Score(Lbase) in this scenario (S'(L0) > S'(Lbase)), it required the following two small tweaks. I think they fix a case where Lbase has a lot of deletions and wonder if you think they make sense.

* `total_downcompact_bytes` uses S'(Li). For all levels this changes it to use compensated file size instead of actual file size. For L0 this additionally changes it to subtract max_bytes_for_level_base.
* `base_level_size` uses S'(Lbase) instead of S*(Lbase). S*(Li) is sum of all compensated file sizes on level i. This change removes files undergoing compaction from `base_level_size`